nexus/src/app/background/blueprint_execution.rs

```diff
@@ -0,0 +1,407 @@
+        opctx: &OpContext,
+        blueprint: &std::sync::Arc<Blueprint>,
+    ) -> Result<(), Vec<anyhow::Error>> {
+        let log = opctx.log.new(o!("comment" => blueprint.comment.clone()));
+        self.deploy_zones(&log, opctx, &blueprint.omicron_zones).await
+    }
+
+    async fn deploy_zones(
+        &self,
+        log: &Logger,
+        opctx: &OpContext,
+        zones: &BTreeMap<Uuid, OmicronZonesConfig>,
+    ) -> Result<(), Vec<anyhow::Error>> {
+        let errors: Vec<_> = stream::iter(zones.clone())
+            .filter_map(|(sled_id, config)| async move {
+                let client = match self.sled_client(&opctx, &sled_id).await {
+                    Ok(client) => client,
+                    Err(err) => {
+                        warn!(log, "{err:#}");
+                        return Some(err);
+                    }
+                };
+                let result = client
+                    .omicron_zones_put(&config)
+                    .await
+                    .with_context(|| {
+                        format!("Failed to put {config:#?} to sled {sled_id}")
+                    });
+
+                match result {
+                    Err(error) => {
+                        warn!(log, "{error:#}");
+                        Some(error)
+                    }
+                    Ok(_) => {
+                        info!(
+                            log,
+                            "Successfully deployed zones for sled agent";
+                            "sled_id" => %sled_id,
+                            "generation" => config.generation.to_string()
+                        );
+                        None
+                    }
+                }
+            })
+            .collect()
+            .await;
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    }
+}
+
+impl BackgroundTask for BlueprintExecutor {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            // Get the latest blueprint, cloning to prevent holding a read lock
+            // on the watch.
+            let blueprint = self.rx_blueprint.borrow_and_update().clone();
+
+            let Some(blueprint) = blueprint else {
+                warn!(&opctx.log,
+                      "Plan execution: skipped";
+                      "reason" => "no blueprint");
+                return json!({"error": "no blueprint" });
+            };
+
+            let result = self.realize_blueprint(opctx, &blueprint).await;
+
+            // Return the result as a `serde_json::Value`
+            match result {
+                Ok(()) => json!({}),
+                Err(errors) => {
+                    let errors: Vec<_> = errors
+                        .into_iter()
+                        .map(|e| format!("{:#}", e))
+                        .collect();
+                    json!({"errors": errors})
```

Comment:
I think it'd be useful to report which blueprint we tried to execute.

Reply:
Done in 596c9be466c0ba54c41201edf0c18d9b8c6c6280